### PR TITLE
Improve exception messages and simplify LayoutPath

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to improve the exception messages in the internal `LayoutPath` class and also do some code simplifications.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign.git pull/886/head:pull/886` \
`$ git checkout pull/886`

Update a local copy of the PR: \
`$ git checkout pull/886` \
`$ git pull https://git.openjdk.org/panama-foreign.git pull/886/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 886`

View PR using the GUI difftool: \
`$ git pr show -t 886`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/886.diff">https://git.openjdk.org/panama-foreign/pull/886.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/panama-foreign/pull/886#issuecomment-1722975849)

### CODE REVIEW

**Review by JornVernee — COMMENTED** [review]
I like the overall idea. I think improving the error messages can help tell a user from _where_ in a layout path an error occurs.

But, currently you are only printing out the inner-most enclosing layout. Another idea would be to print a kind of back trace, where each `enclosing` layout path is listed in the exception message, until we get to the root. That should, I think, give the full image of where in the layout path the error occurs.

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 1]
```diff
@@ -314,15 +320,24 @@ private static LayoutPath derefPath(MemoryLayout layout, MethodHandle handle, La
...
 
-    private void check(Class<?> layoutClass, String msg) {
+    private <T extends MemoryLayout> T requireLayoutType(Class<T> layoutClass) {
         if (!layoutClass.isAssignableFrom(layout.getClass())) {
-            throw badLayoutPath(msg);
+            var name = layoutClass.getSimpleName();
+            var type = name
+                    // Take what is before "Layout"
+                    .substring(0, name.indexOf("Layout"))
+                    .toLowerCase(Locale.ROOT);
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:329]: I'd much prefer the simplicity of just passing in the name of the type as a `String` parameter.

Also, it seems like this wouldn't work for `AddressLayout` ?: `"attempting to select a address element from a non-address layout ..."`

Maybe a couple of simple `checkSequenceLayout`, `checkGroupLayout` would be better to avoid the existing duplication. I kinda feel like the new solution ends up introducing more complexity than it resolves.

**Review by JornVernee — COMMENTED** [review]
Thanks for adding the breadcrumbs, I think they could be useful for debugging.

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 2]
```diff
@@ -314,15 +319,27 @@ private static LayoutPath derefPath(MemoryLayout layout, MethodHandle handle, La
...
+
+    private GroupLayout requireGroupLayout() {
+        return requireLayoutType(GroupLayout.class, "group");
+    }
+
+    private <T extends MemoryLayout> T requireLayoutType(Class<T> layoutClass, String name) {
         if (!layoutClass.isAssignableFrom(layout.getClass())) {
-            throw badLayoutPath(msg);
+            throw badLayoutPath(
+                    String.format("attempting to select a %s element from a non-%s layout: %s (%s)",
```
Comment (minborg) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:333]: While it is good to see the breadcrumbs, having them enclosed in parentheses creates confusion as the layouts themselves contain parentheses as shown in the example below.

Is there a better way (e.g. using `{}`)?

```
Bad layout path: attempting to select a sequence element from a non-sequence layout: [i4(3a)i4(3b)](2) ([[2:[i4(3a)i4(3b)](2)](1)](0) -> [2:[i4(3a)i4(3b)](2)](1) -> [i4(3a)i4(3b)](2))
```

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:333]: Maybe instead of parenthesis, you could have something like : `path =` in the message? Or maybe something like
```
Bad layout path: attempting to select a sequence element from a non-sequence layout: [i4(3a)i4(3b)](2), selected from: [2:[i4(3a)i4(3b)](2)](1), selected from: [[2:[i4(3a)i4(3b)](2)](1)](0)
```

**Review by JornVernee — APPROVED** [review]

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 package jdk.internal.foreign;
 
 import jdk.internal.vm.annotation.ForceInline;
+import jdk.internal.vm.annotation.Stable;
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:29]: This doesn't seem used?

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 4]
```diff
@@ -40,9 +41,13 @@
 import java.lang.invoke.VarHandle;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Locale;
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:44]: This also seems unused?

**Comment by GavinRay97** [comment]
Could I make a suggestion?

For the following `MemoryLayout`:

```java
        MemoryLayout struct = MemoryLayout.structLayout(
                MemoryLayout.sequenceLayout(2, MemoryLayout.structLayout(
                                JAVA_INT.withOrder(ByteOrder.LITTLE_ENDIAN).withName("3a"),
                                JAVA_INT.withOrder(ByteOrder.LITTLE_ENDIAN).withName("3b")
                        ).withName("2")
                ).withName("1")
        ).withName("0");
```

I find the notation like:
```
                "[i4(3a)i4(3b)](2), selected from: " +
                "[2:[i4(3a)i4(3b)](2)](1), selected from: " +
                "[[2:[i4(3a)i4(3b)](2)](1)](0)";
```

Difficult to interpret. Maybe something like the below could be clearer/easier-to-read?
```
(0: struct)
├── (1: sequence) x 2
│   ├── (2: struct)      ERROR: Invalid sequence selection!
│   │   ├── (3a: i4 LE)
│   │   └── (3b: i4 LE)
```

**Comment by mcimadamore** [comment]
> Could I make a suggestion?
> 
> For the following `MemoryLayout`:
> 
> ```java
>         MemoryLayout struct = MemoryLayout.structLayout(
>                 MemoryLayout.sequenceLayout(2, MemoryLayout.structLayout(
>                                 JAVA_INT.withOrder(ByteOrder.LITTLE_ENDIAN).withName("3a"),
>                                 JAVA_INT.withOrder(ByteOrder.LITTLE_ENDIAN).withName("3b")
>                         ).withName("2")
>                 ).withName("1")
>         ).withName("0");
> ```
> 
> I find the notation like:
> 
> ```
>                 "[i4(3a)i4(3b)](2), selected from: " +
>                 "[2:[i4(3a)i4(3b)](2)](1), selected from: " +
>                 "[[2:[i4(3a)i4(3b)](2)](1)](0)";
> ```
> 
> Difficult to interpret. Maybe something like the below could be clearer/easier-to-read?
> 
> ```
> (0: struct)
> ├── (1: sequence) x 2
> │   ├── (2: struct)      ERROR: Invalid sequence selection!
> │   │   ├── (3a: i4 LE)
> │   │   └── (3b: i4 LE)
> ```

Honestly I find the notation hard to parse too - but, I have to admit, I wonder if a structural representation, such as the one in this PR, or the one in the comment above really "fit" an exception message - they seem more akin to compiler messages to me.
